src/processing/strategies/transform.rs

```diff
@@ -0,0 +1,37 @@
+use crate::processing::strategies::{
+    CommitRequest, InvalidMessage, MessageRejected, ProcessingStrategy,
+};
+use crate::types::Message;
+
+pub struct Transform<TPayload: Clone + Send + Sync, TTransformed: Clone + Send + Sync> {
+    pub function: fn(Message<TPayload>) -> Result<TTransformed, InvalidMessage>,
+    pub next_step: Box<dyn ProcessingStrategy<TTransformed>>,
+}
+
+impl<TPayload: Clone + Send + Sync, TTransformed: Clone + Send + Sync> ProcessingStrategy<TPayload>
+    for Transform<TPayload, TTransformed>
+{
+    fn poll(&mut self) -> Option<CommitRequest> {
+        None
+    }
+
+    fn submit(&mut self, message: Message<TPayload>) -> Result<(), MessageRejected> {
+        // TODO: Handle InvalidMessage
+        let transformed = (self.function)(message.clone()).unwrap();
```

Comment:
Have you considered passing the message by reference? If a processing method decides it is a good idea to update the message passed in input they deserve the dire consequences. Passing the message by reference would actually force us to write sound processing methods if anything as bad code would not compile.

Reply:
Yes, I think we should. I wasn't going to tackle that in this PR but on second thoughts I could probably quickly adjust it.

Reply:
Ok, updated to pass reference to message. We still need to revisit some of the other types we are passing around later.

Reply:
Ok, after giving this a bit more thought, I think there are also some downsides of not having each strategy own the message it is processing. I think we should consider this a bit more before making this change so I would not do this as part of this PR.

Reply:
In the current scenario, the message is owned by the submit method. Why not just extract the contents like `partition`, `payload` and `offset` into local variables on the stack before calling `self.function`. They should be pretty small in size. And then pass ownership of the `message` to `self.function`? Then you wouldn't require to clone it right?